java/src/main/java/com/google/appengine/tools/pipeline/impl/PipelineManager.java

```diff
@@ -413,11 +414,74 @@ public static void deletePipelineRecords(String pipelineHandle, boolean force, b
     backEnd.deletePipeline(key, force, async);
   }
 
+  /**
+   * Fills a the {@link PromisedValue}.
+   * @param promiseHandle the key to the promised value slot.
+   * @param value the value to fill.
+   * @throws NoSuchObjectException If there is no Job with the given key.
+   * @throws OrphanedObjectException If the slot has been orphaned.
+   */
   public static void acceptPromisedValue(String promiseHandle, Object value)
       throws NoSuchObjectException, OrphanedObjectException {
+    Slot slot = queryPromisedValueSlot(promiseHandle);
+    JobRecord generatorJob = getGeneratorJob(slot);
+    UpdateSpec updateSpec = new UpdateSpec(slot.getRootJobKey());
+    registerSlotFilled(updateSpec, generatorJob.getQueueSettings(), slot, value);
+    backEnd.save(updateSpec, generatorJob.getQueueSettings());
+  }
+  
+  private static JobRecord getGeneratorJob(Slot slot)
+      throws OrphanedObjectException, NoSuchObjectException {
+    Key generatorJobKey = slot.getGeneratorJobKey();
+    if (null == generatorJobKey) {
+      throw new RuntimeException(
+          "Pipeline is fatally corrupted. Slot for promised value has no generatorJobKey: " + slot);
+    }
+    JobRecord generatorJob = backEnd.queryJob(generatorJobKey, JobRecord.InflationType.NONE);
+    if (null == generatorJob) {
+      throw new RuntimeException("Pipeline is fatally corrupted. "
+          + "The generator job for a promised value slot was not found: " + generatorJobKey);
+    }
+    String childGraphGuid = generatorJob.getChildGraphGuid();
+    if (null == childGraphGuid) {
+      // The generator job has not been saved with a childGraphGuid yet. This can happen if the
+      // promise handle leaked out to an external thread before the job that generated it
+      // had finished.
+      throw new NoSuchObjectException(
+          "The framework is not ready to accept the promised value yet. "
+          + "Please try again after the job that generated the promis handle has completed.");
+    }
+    if (!childGraphGuid.equals(slot.getGraphGuid())) {
+      // The slot has been orphaned
+      throw new OrphanedObjectException(KeyFactory.keyToString(slot.getKey()));
+    }
+    
+    return generatorJob;
+  }
+
+  /**
+   * Get the {@link Slot} for a promise handle that can be used to construct a
+   * {@link PromisedValue} for use in child {@link Job}s. It will attempt to lookup
+   * the key 5 times with an exponential back-off just in case it has been requested
+   * before the promise has been registered. 
+   * @param promiseHandle Key to find the {@link Slot}
+   * @return A {@link Slot} if the handle can be resolved otherwise <code>null</code>.
+   */
+  public static Slot getPromisedValueSlot(String promiseHandle) {
```

Comment:
If you are worried about masking the exception I can follow the example of acceptPromisedValue and throw NoSuchObjectException and OrphanedObjectException, but the method itself is required because queryPromisedValueSlot is private and I am not sure I should change it's scope.